crates/flipperzero/src/furi/log/mod.rs

```diff
@@ -0,0 +1,166 @@
+//! Furi Logging system.
+
+mod metadata;
+pub use metadata::{Level, LevelFilter};
+
+/// The standard logging macro.
+///
+/// This macro will generically log with the specified `Level` and C-style format argument
+/// list.
+///
+/// # Examples
+///
+/// ```
+/// use flipperzero::{error, furi::log::Level};
+///
+/// # fn main() {
+/// let error_code = 42;
+/// log!(Level::ERROR, "Failed to handle the florp: %d", error_code);
+/// log!(target: "events", Level::INFO, "Finished the documentation!");
+/// # }
+/// ```
+#[macro_export]
+macro_rules! log {
+    (target: $target:expr, $lvl:expr, $msg:expr $(, $arg:expr)*) => ({
+        if $lvl <= $crate::furi::log::LevelFilter::current() {
+            unsafe {
+                $crate::__macro_support::__sys::furi_log_print_format(
+                    $crate::__macro_support::__level_to_furi($lvl),
+                    $crate::__macro_support::__sys::c_string!($target),
+                    $crate::__macro_support::__sys::c_string!($msg),
```

Comment:
These force `$target` and `$msg` to be literals, though not specifically string literals (instead we get the string representation of whatever literal is passed in, with a trailing null byte).

Reply:
We now use `ufmt::uwrite!` to constrain `$msg` instead.